crate/operator/main.py

```diff
@@ -382,129 +382,169 @@ async def cluster_create(
+            do_restart = True
         elif field_path == ("spec", "nodes", "master", "replicas"):
-            do_scale_master = True
-            scale_master_diff_item = kopf.DiffItem(
-                operation, field_path, old_value, new_value
-            )
+            do_scale = True
         elif field_path == ("spec", "nodes", "data"):
-            # TODO: check for data node order, added or removed types, ...
-            if len(old_value) != len(new_value):
-                raise kopf.PermanentError(
-                    "Cannot handle changes to the number of node specs."
-                )
-            scale_data_diff_items = []
-            for node_spec_idx in range(len(old_value)):
-                old_spec = old_value[node_spec_idx]
-                new_spec = new_value[node_spec_idx]
-                inner_diff = calc_diff(old_spec, new_spec)
-                for (
-                    inner_operation,
-                    inner_field_path,
-                    inner_old_value,
-                    inner_new_value,
-                ) in inner_diff:
-                    if inner_field_path == ("replicas",):
-                        do_scale_data = True
-                        scale_data_diff_items.append(
-                            kopf.DiffItem(
-                                inner_operation,
-                                (str(node_spec_idx),) + inner_field_path,
-                                inner_old_value,
-                                inner_new_value,
-                            )
-                        )
-                    else:
-                        logger.info(
-                            "Ignoring operation %s on field %s",
-                            operation,
-                            field_path + (str(node_spec_idx),) + inner_field_path,
-                        )
-        else:
-            logger.info("Ignoring operation %s on field %s", operation, field_path)
-
-    async with ApiClient() as api_client:
-        apps = AppsV1Api(api_client)
-        coapi = CustomObjectsApi(api_client)
-        core = CoreV1Api(api_client)
-
-        if do_upgrade:
-            webhook_upgrade_payload = WebhookUpgradePayload(
+            do_scale = True
+
+    if do_upgrade:
+        if new_cycle:
+            context.state_machine.add(State.UPGRADE)
+        kopf.register(fn=UpgradeSubHandler(context)(), id="upgrade")
+
+    if do_restart:
+        if new_cycle:
+            context.state_machine.add(State.RESTART)
+        kopf.register(
+            fn=RestartSubHandler(context)(),
+            id="restart",
+            timeout=config.ROLLING_RESTART_TIMEOUT,
+        )
+
+    if do_scale:
+        if new_cycle:
+            context.state_machine.add(State.SCALE)
+        kopf.register(
+            fn=ScaleSubHandler(context)(), id="scale", timeout=config.SCALING_TIMEOUT
+        )
+
+    patch.status["updateContext"] = context.serialize()
+
+
+class UpgradeSubHandler(StateBasedSubHandler):
```

Comment:
I think moving those sub-handlers into a separate file (or separate files) would make sense to keep the main clean with just the handler invocations

Reply:
Good idea!